### ISSUE TITLE
[Vim] Files disappear on write

### ISSUE THREAD

**Issue by gigglemoose** [srid/emanote#180]
Emanote loses files when edited through vim on my setup.

- EndeavourOS (Arch-based distro) running dwm and st
- Vim text editor
- Vivaldi browser

Everything at its latest available version.

Screencast displaying bug: https://www.youtube.com/watch?v=WyTk73EaCS0 
Script file: https://pastebin.com/g45qapn6
Timing file: https://pastebin.com/Y3kAJJEM

I create a note directory ~/emanote, add an index.md, show it working, edit it, and Emanote reverts to the default index page. Then I create a ~/emanote/Math/file.md, show it working, edit it, and it disappears as well. I also show that killing and restarting the server restores the pages.

**Comment by srid** [comment]
I looked at the YouTube video. The log on the left is very telling. When you save a file, your text editor sends two filesystem events to Emanote

- "Removed index.md"
- "Added 4913"

Normally one would see the following two events instead if the text editor saves the file by deleting it first (as strange a behaviour as that is):

- "Removed index.md"
- "Added index.md"

Or, the typical scenario is:

- "Modified index.md"

So the question remains - what's unique about your text editor that causes this behaviour?

Does this happen only with your Vim text editor, or with other editors as well? Do other editors as well create this strange "4913" file?

**Comment by gigglemoose** [comment]
Alright, so I've confirmed that this is just a weird Vim thing. Vim in nocompatibility mode (which is what a large majority of vim users will be in because it's actually the default if a user has any .vimrc file at all) has atypical behaviour in how it actually saves files. Anybody who cares about the details of why this is happening can check out [this link](https://vi.stackexchange.com/questions/4038/why-does-set-nocompatible-result-in-vim-saving-extra-all-numeric-temporary-fi) which has pretty much all of the info needed to understand the problem.

The tl;dr for solving this problem as a user is to `set backupcopy=yes` in your .vimrc file. It's essential to do this *after* you `set nocompatible`, if you're the type of person to add that to be explicit. 

It's unclear whether or not that default behaviour is worth the effort of supporting from a development perspective, but this workaround should help fellow Vim users not go any more insane than they already are.

**Comment by j-steinbach** [comment]
Question: Does this apply for Neovim as well? I am not sure if you are talking about _Vim_ specifically or mean the whole range of _vi_, _vim_ and _neovim_.

**Comment by gigglemoose** [comment]
I'm most familiar with Vim, but my testing seems to indicate that neovim has the same problem (and solution) and vi actually works with Emanote out of the box

**Comment by j-steinbach** [comment]
Thank you for clarifying that. I run (and am new to) Neovin, and just checked -- I think I don't yet have `nocompatible` set yet, so your findings come at a good time for me :+1: 

---

I think we could create a general "Setting up Vim for Emanote" discussion. (The WIP Emanote docs only describe a VS Code setup, whereas the old Neuron docs mainly lists a few [plugins](https://neuron.zettel.page/editor)).

**Comment by srid** [comment]
We can create a dedicated page for Vim? Similar to https://emanote.srid.ca/resources/vscode

And then describe this issue in there for those interested.

PR for that is welcome! I'll leave this open for a few days, but it doesn't look like there is any other actionable thing here.

**Comment by waynr** [comment]
> We can create a dedicated page for Vim? 

This would definitely help me right now :). I'm just learning about zettelkasten and have started to use `vim-zettel`/`vimwiki` which so far falls short in a few ways.

**Comment by j-steinbach** [comment]
I am still in the midst of learning/configuring vim, but I think either [ZK](https://github.com/mickael-menu/zk) or [Telekasten](https://github.com/renerocksai/telekasten.nvim) is the way to go. (Don't yet have them set up)

---

E: Maybe you could say how `vim-zettel` and `vimwiki` fall short? I.e. what are they missing?

E2: @srid can you update your previous comment with a working url? I think it is https://emanote.srid.ca/resources/vscode now

E3: I've started a general disucssion in https://github.com/srid/emanote/discussions/237

**Comment by srid** [comment]
@j-steinbach I don't use Vim myself for note-taking, but seeing as you have been exploring this space (#237), I trust you with whatever Vim recommendation you come up with in the docs. If you could write up a draft (or open a PR), I'd be happy to get it added to the documentation!

**Comment by j-steinbach** [comment]
@srid Here is a quick small draft. I will try to find the time to improve it, but I am unfortunately quite busy (just moved to a new country).


# Neovim for note-taking

> A quick guide for setting up [Neovim](http://neovim.io/) to take notes.

_Why Neovim?_ Because VIM simply is the best editing paradigm for working with text. You go VIM and you never go back. (Some people however move forwards to [Doom Emacs](https://github.com/doomemacs/doomemacs) or [kakoune](http://kakoune.org/)).. :).

_How do I learn VIM-keybinds?_ To learn the keybinds, simply open Neovim and type `:Tutor` to start the interactive vimtutor. You can also get help with `:help vimtutor`. For a more in-depth guide see [iggredible/Learn-Vim](https://github.com/iggredible/Learn-Vim). Note: Neovim can be confusing at first, so watching a few videos also helps :)

_What is the difference between VIM and Neovim?_ Neovim is VIM, just better. Modern, sane defaults, community-based development cycle and nearly fully configurable in Lua.

## Where to start

A. _These VIM-style keybinds sound interesting, but I really don't want to get my hands dirty. Isn't there an easier way?._ Have a look at [VS Code](https://code.visualstudio.com/), [Obsidian](https://obsidian.md/) or [logseq](https://logseq.com/). They can all be configured to use VIM-keybinds.

B. _I want to use Neovim, but I don't know where to start. I just want to get note-taking as soon as possible._ If you don't know what you are doing, why don't you start with a curated base config, for example [Spacevim](https://spacevim.org/) (or [liuchengxu/spacevim](https://github.com/liuchengxu/space-vim))? Then add one of the different wiki plugins on top: [Vimwiki](https://github.com/vimwiki/vimwiki), [wiki.vim](https://github.com/lervag/wiki.vim) or [Telekasten](https://github.com/renerocksai/telekasten.nvim). Then slowly go from there (See C).

C. _I want to get my hands dirty. I have lots of time, energy and motivation and really feel like learning and mastering a new editor/craft._ Excellent choice! Take a look at the plugins and resources in #237. Make sure to read the [nanotee/nvim-lua-guide](https://github.com/nanotee/nvim-lua-guide). Watch videos, join the [Neovim Discourse](https://neovim.discourse.group/), search for help online.

D. _I know what I am doing. I am a Neovim wizard!_ Then why are you reading this? Oh, wait, you want to help improve this guide? Just submit a PR **here** (TODO)

Bonus: For coolness you can also use a Neovim GUI like [Neovide](https://github.com/neovide/neovide).

Neovim will feel strange at first, but it is very rewarding once you grasp the "new" way of editing text. Just keep in mind that it is all about comfortably writing your notes. Making everything more complicated is easy, but having a simple working workflow is hard.
#

**Comment by applejag** [comment]
For reference, a docs page for this issue has been published at https://emanote.srid.ca/resources/vim#known-issue-glitches-when-saving-file via #319.

Sorry @j-steinbach but I skipped your "background about Vim" docs as I thought it was a little out of scope for Emanote to have docs about. If you insist then you're probably free to open PR on it

**Comment by j-steinbach** [comment]
@jilleJr Np, people will find it in the issues if they search for it (And you even linked to the discussion :+1: ).
